mkosi/__init__.py

```diff
@@ -4277,13 +4277,8 @@ def install_unified_kernel(
         for kver, kimg in gen_kernel_images(args, root):
             if args.image_id:
                 image_id = args.image_id
-                if args.image_version:
-                    partlabel = f"{args.image_id}_{args.image_version}"
-                else:
-                    partlabel = f"{args.image_id}"
```

Comment:
first i want to say that i didn't know there was other tests than the ones ran by pytest (the integration tests in the ci), so i will see what makes those fail and try to fix that.
i assume that the ci tests that failed probably failed because they broke "building different image versions".


i made that change because:
when building images with no modifications to the production code of mkosi,
when building images that have luks root partition and an explicitely set image id, the `root=PARTLABEL=IMAGE_ID` kernel parameter is sent to dracut, and booting the image fails;
when building images that have luks root partition and **no** explicitely set image id, the `root` kernel parameter is **not** sent to dracut, and booting the image works.

when the image fails to boot, the error is that it tried to mount the partition with the label provided in the `root` kernel parameter without decrypting it first.

when the image boots fine, no `root` kernel parameter is sent to dracut, and i could not (and still can't) see any place where information is passed to the kernel or initrd to specify which partition to boot.
so i assume that dracut, or whatever process is ran inside the initrd must autodectect what partition to boot and correctly ask for luks password (and decrypt the partition) when it sees a luks partition.

i tried to quickly read some documentation on dracut and systemd, but i could not see how it seems to be able to find the right partition (and to ask password and decrypt in the case of luks) when there is no kernel paremeter passed to specify which partition to boot.

Reply:
No need to apologise! Thanks for the analysis of the problem.

I'm unsure that the failing tests have nothing to do with your current change here, the comment about building different image versions is regarding that this would break building system extension images (see `man systemd-systext`).

Without any kernel parameters the discoverable partition scheme should (and apparently does) pick up the right partitions to mount, but in apparently we have some ill-defined mixture here. Will have to look more closely at this.

Reply:
by "discoverable partition scheme" do you mean [this](https://systemd.io/DISCOVERABLE_PARTITIONS/) ? that's relly cool i didn't know about this.

i read the [systemd-sysext man page](https://man.archlinux.org/man/systemd-sysext.8) but i still don't understand how not passing the root kernel parameter could break that. it looks like to me that it's either finding the extension via discoverable partition, in which case it does not need the kernel parameter, or it reads the file from already mounted fs, in which case you's have to already have mounted the fs in another way, which i am assuming would be fstab or discoverable partitions, or means custom change made by a user that is not made by mkosi.
maybe you could explain to me more how you think my change would break this.

also, now that i know about discoverable partitions, if mkosi always generate an image with discoverable partitions, then i don't see any need for the root kernel prameter.
and the `partlabel` variable which i have removed was **only** (afaict) used to set the root kernel parameter.

Reply:
@nlacombe42 Sorry for the late reply, I was travelling last week. 

We can't do without the partlabel because we document that it will be set when `ImageID` is set and it's documented to be used in the discoverable partition scheme as being used for version comparisons if multiple images with the same image ID are being passed. The relevant PR introducing this is #754.

Also thinking about the discoverable partition scheme was a bit of a red herring here (sorry), since setting `root=` on the kernel cmdline disables the generator.

Reply:
Euhh I'm slightly confused here as well on why we configure PARTLABEL on the cmdline explicitly if we're using image IDs. Don't we always want to use systemd-gpt-generator to figure out the partition to mount? It should automatically use the partition label in its version comparison logic to figure out the right partition no?

cc @poettering